File: 023-merge-k-sorted-lists/merge-k-sorted-lists.py
# -*- coding:utf-8 -*-


# Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.
#
# Example:
#
#
# Input:
# [
# &nbsp; 1-&gt;4-&gt;5,
# &nbsp; 1-&gt;3-&gt;4,
# &nbsp; 2-&gt;6
# ]
# Output: 1-&gt;1-&gt;2-&gt;3-&gt;4-&gt;4-&gt;5-&gt;6
#
#


# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def mergeKLists(self, lists):
        
        if not lists:
            return None

        dummyNode = cur = ListNode(0)
        minHeap = [(l.val, l) for l in lists if l]
        heapq.heapify(minHeap)

        while minHeap:
            cur.next = heapq.heappop(minHeap)[1]
            cur = cur.next

            if cur.next:
                heapq.heappush(minHeap, (cur.next.val, cur.next))

        return dummyNode.next

File: 023-merge-k-sorted-lists/merge-k-sorted-lists.cpp
// Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.
//
// Example:
//
//
// Input:
// [
//   1->4->5,
//   1->3->4,
//   2->6
// ]
// Output: 1->1->2->3->4->4->5->6
//
//


/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) return nullptr;
        int j = lists.size() - 1;
        while (j > 0) {
            for (int i = 0; i < j; ++i, --j) {
                lists[i] = merge2(lists[i], lists[j]);
            }
        }
        return lists[0];
    }
    ListNode* merge2(ListNode* a, ListNode* b) {
        ListNode dummy(0);
        auto p = &dummy;
        while (a || b) {
            if (!b || (a && b && a->val < b->val)) {
                p->next = a;
                a = a->next;
            } else {
                p->next = b;
                b = b->next;
            }
            p = p->next;
        }
        return dummy.next;
    }
};
